lib/wifi/src/ap.cpp

```diff
@@ -50,20 +63,44 @@ ap_status_t AccessPoint::getStatus() {
   return this->status;
 }
 
-void AccessPoint::handleIncomingDataPacket(wifi_promiscuous_pkt_t* packet) {
+void AccessPoint::receive(wifi_promiscuous_pkt_t* packet) {
+  // update counters
   status.rxFrames++;
   if (packet->rx_ctrl.rx_state != 0) {
     status.rxErrors++;
     return;
   }
 
-  status.rssi = EXPONENTIAL_SMOOTHING_ALPHA * packet->rx_ctrl.rssi +
-                (1 - EXPONENTIAL_SMOOTHING_ALPHA) * status.rssi;
-  status.length = EXPONENTIAL_SMOOTHING_ALPHA * packet->rx_ctrl.sig_len +
-                  (1 - EXPONENTIAL_SMOOTHING_ALPHA) * status.length;
+  exponentialSmoothing<int16_t>(status.rssi, packet->rx_ctrl.rssi);
+  exponentialSmoothing<uint16_t>(status.length, packet->rx_ctrl.sig_len);
 
+  // get layer 2 data
   auto l2data = getLayer2Data(packet->payload, packet->rx_ctrl.sig_len);
-  // TODO send to a queue...
+  if (l2data.length <= 0)
+    return;
+
+  // alocating memory for the data to be read elsewhere
+  void* payload = malloc(l2data.length);
+  if (payload == nullptr) {
+    status.rxErrors++;
+    return;
+  }
+  memcpy(payload, l2data.payload, l2data.length);
+  l2data.payload = payload;
+
+  // send to the queue
+  rxQueue->push(&l2data);
+}
+
+void AccessPoint::transmit() {
+  auto layer2Data = txQueue->poll();
+  if (layer2Data == nullptr)
+    return;
+
+  // send over 802.11
+
+  free(layer2Data->payload);
+  free(layer2Data);
```

Comment:
don't need to free this data here, unless you're planning to malloc first

Reply:
Yes, you're right... This code is actually unreachable for now, since the LoRa receiver should malloc and write some data into the queue. I'm gonna put some better explanation on the TODO above.